azkaban-common/src/main/java/azkaban/project/JdbcProjectImpl.java

```diff
@@ -968,4 +972,75 @@ public void cleanOlderProjectVersion(final int projectId, final int version)
       throw new ProjectManagerException("clean older project transaction failed", e);
     }
   }
+
+  @Override
+  public void uploadFlowFile(final int projectId, final int projectVersion, final int flowVersion,
+      final File flowFile) throws ProjectManagerException {
+    logger.info(String
+        .format(
+            "Uploading flow file %s, version %d for project %d, version %d, file length is [%d bytes]",
+            flowFile.getName(), flowVersion, projectId, projectVersion, flowFile.length()));
+
+    if (flowFile.length() > MAX_FLOW_FILE_SIZE_IN_BYTES) {
+      throw new ProjectManagerException("Flow file length exceeds 10 MB limit.");
+    }
+
+    final byte[] buffer = new byte[MAX_FLOW_FILE_SIZE_IN_BYTES];
+    final String INSERT_FLOW_FILES =
+        "INSERT INTO project_flow_files (project_id, project_version, flow_name, flow_version, "
+            + "modified_time, "
+            + "flow_file) values (?,?,?,?,?,?)";
+
+    try (final FileInputStream input = new FileInputStream(flowFile);
+        final BufferedInputStream bufferedStream = new BufferedInputStream(input)) {
+      final int size = bufferedStream.read(buffer);
+      logger.info("Read bytes for " + flowFile.getName() + ", size:" + size);
+      final byte[] buf = Arrays.copyOfRange(buffer, 0, size);
+      try {
+        this.dbOperator
+            .update(INSERT_FLOW_FILES, projectId, projectVersion, flowFile.getName(), flowVersion,
+                System.currentTimeMillis(), buf);
+      } catch (final SQLException e) {
+        throw new ProjectManagerException(
+            "Error uploading flow file " + flowFile.getName() + ", version " + flowVersion + ".",
+            e);
+      }
+    } catch (final IOException e) {
+      throw new ProjectManagerException(
+          String.format(
+              "Error reading flow file %s, version: %d, length: [%d bytes].",
+              flowFile.getName(), flowVersion, flowFile.length()));
+    }
+  }
+
+  @Override
+  public File getUploadedFlowFile(final int projectId, final int projectVersion,
```

Comment:
what do you think of the argument order like this (projectId, projectVersion, flowName, flowVersion)? since It's good to keep the consistency of argument order(id, version, id, version).

Reply:
FlowName actually refers to the flow file name, not the flowId. The name seems a little confusing here.